pyomo/contrib/alternative_solutions/obbt.py

```diff
@@ -0,0 +1,370 @@
+            constraint will not be added to the model.
+        refine_discrete_bounds : boolean
+            Boolean indicating that new constraints should be added to the
+            model at each iteration to tighten the bounds for discrete
+            variables.
+        warmstart : boolean
+            Boolean indicating that the solver should be warmstarted from the
+            best previously discovered solution.
+        solver : string
+            The solver to be used.
+        solver_options : dict
+            Solver option-value pairs to be passed to the solver.
+        tee : boolean
+            Boolean indicating that the solver output should be displayed.
+        quiet : boolean
+            Boolean indicating whether to suppress all output.
+
+        Returns
+        -------
+        variable_ranges
+            A Pyomo ComponentMap containing the bounds for each variable.
+            {variable: (lower_bound, upper_bound)}. An exception is raised when
+            the solver encountered an issue.
+        solutions
+            [Solution]
+    """
+
+    # TODO - parallelization
+
+    if not quiet:  # pragma: no cover
+        print("STARTING OBBT ANALYSIS")
+
+    if warmstart:
+        assert (
+            variables == "all"
+        ), "Cannot restrict variable list when warmstart is specified"
+    all_variables = aos_utils.get_model_variables(model, "all", include_fixed=False)
+    if variables == "all":
+        variable_list = all_variables
+    else:
+        variable_list = list(variables)
+    if warmstart:
+        solutions = pe.ComponentMap()
+        for var in all_variables:
+            solutions[var] = []
+
+    num_vars = len(variable_list)
+    if not quiet:  # pragma: no cover
+        print(
+            "Analyzing {} variables ({} total solves).".format(num_vars, 2 * num_vars)
+        )
+    orig_objective = aos_utils.get_active_objective(model)
+
+    use_appsi = False
+    if "appsi" in solver:
+        opt = appsi.solvers.Gurobi()
+        for parameter, value in solver_options.items():
+            opt.gurobi_options[parameter] = value
+        opt.config.stream_solver = tee
+        opt.config.load_solution = False
+        results = opt.solve(model)
+        condition = results.termination_condition
+        optimal_tc = appsi.base.TerminationCondition.optimal
+        infeas_or_unbdd_tc = appsi.base.TerminationCondition.infeasibleOrUnbounded
+        unbdd_tc = appsi.base.TerminationCondition.unbounded
+        use_appsi = True
+    else:
+        opt = pe.SolverFactory(solver)
+        for parameter, value in solver_options.items():
+            opt.options[parameter] = value
+        try:
+            results = opt.solve(
+                model, warmstart=warmstart, tee=tee, load_solutions=False
+            )
+        except:
+            # Assume that we failed b.c. of warm starts
+            results = None
```

Comment:
I think the bare `except` here is probably pretty dangerous. Can you specifically catch the warmstart exception and if it's not that, just raise it to be safe?

Reply:
I don't disagree.  But can we trust that solvers will consistently return the same error for failures during a warmstart?  Do we have such an exception defined?

Reply:
@emma58 Do you have a recommendation for how to handling the exception handling differently?

Reply:
I'm actually unclear on when you get an exception because of warmstarts... At least with Gurobi it usually just rejects your warmstart and goes forward. So honestly if it were me I would not catch anything here and assume all exceptions are something the user needs to know. But maybe @jlgearh knows more--I'm not sure what the original reason behind this try-except was?

Reply:
I cleaned this up.  This is catching the case where a solver doesn't implement the 'warmstart' option.